[EA] Add recommendations to the posts page

oetherington opened the pull request:
Add recommendations to the post page. Notable things in this PR:

- We add a whole new recommendations algorithm architecture using "strategies"
- One strategy in particular - the "feature" strategy - is used to implement several of the others using small composable scores
- Adds a /admin/sampleRecommendations page to check the recommendations for the current frontpage
- Adds a script to generate random samples of recommendations
- Make footnotes collapsed by default
- Adds an "exclude from recommendations" option to the triple dot menu on posts for admins and moderators
- Adds general architecture for defining postgres views that are refreshed by a cron job
- Improves behaviour when scolling to comments

<img width="834" alt="Screenshot 2023-05-03 at 16 17 50" src="https://user-images.githubusercontent.com/5075628/235962317-39569875-5415-4762-b1a1-76518090933d.png">

<img width="790" alt="Screenshot 2023-05-03 at 16 17 28" src="https://user-images.githubusercontent.com/5075628/235962313-373f2a2e-1506-45e3-abb0-045ceff573fa.png">

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204461329899219) by [Unito](https://www.unito.io)

jpaddison3 commented:
- [ ] We should exclude from recommendations if author is deleted

jpaddison3 commented on packages/lesswrong/components/posts/PostsPage/PostsPage.tsx:
```diff
@@ -384,9 +384,12 @@ const PostsPage = ({post, refetch, classes}: {
       </div>
 
-      {isEAForum &&
+      {isEAForum && !post.question && !sequenceId &&
```
:thankyou:

Will-Howard commented on packages/lesswrong/components/posts/PostsPage/CollapsedFootnotes.tsx:
```diff
@@ -0,0 +1,137 @@
+    : {preview: "", rest: "", totalCount: 0};
+}
+
+const CollapsedFootnotes = ({
```
1. This isn't working for me, when I load a page with a footnote hash in the url I can briefly see the footnote but then it disappears
2. I would quite like some way of seeing all the footnotes at once, sometimes the footnotes are the bulk of the post

Is this here so that the bottom of the post isn't too busy? I would be ok with merging without it. Usually there aren't that many footnotes, and people will scroll to the comments anyway so they are likely to see the recommendations

oetherington replied:
1. I couldn't reproduce the issue. What post did you try it on - I'm wondering if maybe there are some posts that use differently formatted markup than what I was expecting (for instance, I'd expect [http://localhost:3000/s/isENJuPdB3fhjWYHd/p/7JxsXYDuqnKMqa6Eq](http://localhost:3000/s/isENJuPdB3fhjWYHd/p/7JxsXYDuqnKMqa6Eq) to work correctly).
2. There should be a "show all footnotes" button - maybe it's not there for reasons related to (1)?

Will-Howard replied:
[This](https://beta.effectivealtruism.org/posts/oCeENfefZkAQhEh5b/feature-announcement-rich-text-editor-footnotes#fncys7a23lnfa) is the post I was looking at, which definitely does have the footnotes on [prod](https://forum.effectivealtruism.org/posts/oCeENfefZkAQhEh5b/feature-announcement-rich-text-editor-footnotes#fncys7a23lnfa). I wonder if it's do to with not having enough to make the "show all" appear?

Will-Howard commented on packages/lesswrong/components/recommendations/RecommendationsSamplePage.tsx:
Great addition

Will-Howard commented on packages/lesswrong/lib/collectionIndexUtils.ts:
What are the changes in this file for?

oetherington replied:
Because we needed to add `ensureCustomPgIndex` (ie; the ability to create indexes with custom SQL rather than with Mongo index specifications on collections).

oetherington replied:
(Which is used by the `UniquePostUpvoters` view)

Will-Howard reviewed (approved):
This is incredible 😍! ASTLAC with the only blocker being the footnote comments

oetherington commented:
@Will-Howard Thanks for the review - I'll take a look over your comments shortly.

I also just realised that I never tagged @agnesstenlund in this PR, and it would probably be sensible for her to have a quick look too before we deploy this.

oetherington commented on packages/lesswrong/server/recommendations/BestOfStrategy.ts:
```diff
@@ -0,0 +1,29 @@
+/**
+ * A recommendation strategy that returns the highest voted posts that the user
+ * hasn't viewed. Note that, for performance reasons, the scores are not inflation
+ * adjusted.
```
Potentially. It might not be worth it for this, but we could probably use that same view in frontpage queries. That may or may not give some performance benefit.

oetherington commented on packages/lesswrong/server/recommendations/CollabFilterStrategy.ts:
```diff
@@ -0,0 +1,45 @@
+ * This class implements a simple item-item collaborative filtering algorithm for
+ * post recommendations. It works by storing the set of unique upvoters for each
+ * post in a materialized view called `UniquePostUpvoters`. We then compute the
+ * Jaccard index between the source post and each recommendable target post,
```
Exactly

oetherington commented:
@agnesstenlund Thanks!

I've just fixed the styles for the "load more". The color was actually working correctly, but we needed to set the opactity too. We have an annoying default in the code base where the default hover style for _all_ links sets the opacity to 0.5, which means that if you want to add a custom color you also need to override this at the same time.

There were two reasons I added the ellipsis for recommended posts:

1. If people see a post that looks interesting but don't want to read it straight away they can bookmark it
2. Admins have easy access to the "exclude from recommendations" button if they see a recommendation for a post that shouldn't be recommended

It could be that that's just not important though? Let me know if you still want to remove it - it's a really quick change.

agnesstenlund commented:
thanks a lot @oetherington! 🙏  I think the reasoning for showing the ellipsis makes sense. Not sure exactly how important those things are (2 seems pretty important?), but feels worth keeping and tracking later if we can get rid of it. If we're keeping it I'd love to try showing it within the post item instead, like this:
![Screenshot 2023-05-09 at 10 33 58](https://user-images.githubusercontent.com/13235378/237041163-241677c0-8b56-459c-ae6e-f9b87236cd00.png)

darkruby501 commented:
Neat!!
